autocomplete/complete.fish

```diff
@@ -1,16 +1,48 @@
-complete -c tldr -xf -s v -d "print verbose output"
-complete -c tldr -xf -l version -d "print version and exit"
-complete -c tldr -xf -s h -l help -d "print this help and exit"
-complete -c tldr -xf -s u -l update -d "update local database"
-complete -c tldr -xf -s c -l clear-cache -d "clear local database"
-complete -c tldr -rf -f -s p -l platform -a "linux osx sunos common" -d "select platform, supported are linux / osx / sunos / common"
-complete -c tldr -r -s r -l render -a PATH -d "render a local page for testing purpose"
+function __tldr_not_contain_standalone_opt
+        __fish_not_contain_opt -s v
+    and __fish_not_contain_opt      version
+    and __fish_not_contain_opt -s l list
+    and __fish_not_contain_opt -s h help
+    and __fish_not_contain_opt -s u update
+    and __fish_not_contain_opt -s c clear-cache
+end
+
+function __tldr_no_os_choice_opt
+        __fish_not_contain_opt      linux osx sunos
+    and __fish_not_contain_opt -s r render
+    and __tldr_not_contain_standalone_opt
+end
+
+function __tldr_no_os_choice_opt_and_p
+        __tldr_no_os_choice_opt
+    and __fish_not_contain_opt -s p platform
+end
+
+function __tldr_positional
+        __fish_use_subcommand
+    and __tldr_not_contain_standalone_opt
+end
+
+function __tldr_positional_no_os
+        __tldr_positional
+    and __fish_not_contain_opt      linux osx sunos
+    and __fish_not_contain_opt -s p platform
+end
+
+complete -c tldr -f  -n "__fish_not_contain_opt -s v"  -s v                -d "print verbose output"
+complete -c tldr -f  -n  __fish_is_first_arg                -l version     -d "print version and exit"
+complete -c tldr -f  -n  __fish_is_first_arg           -s l -l list        -d "list all entries in the local database"
+complete -c tldr -f  -n  __fish_is_first_arg           -s h -l help        -d "print help and exit"
+complete -c tldr -f  -n  __fish_is_first_arg           -s u -l update      -d "update local database"
+complete -c tldr -f  -n  __fish_is_first_arg           -s c -l clear-cache -d "clear local database"
+complete -c tldr -x  -n  __tldr_no_os_choice_opt       -s p -l platform    -d "select platform" -a "linux osx sunos common"
+complete -c tldr -f  -n  __tldr_no_os_choice_opt_and_p      -l linux       -d "show command page for Linux"
+complete -c tldr -f  -n  __tldr_no_os_choice_opt_and_p      -l osx         -d "show command page for macOS"
```

Comment:
This doesn't quite work, not completely.

When I type
```
tldr --o
```
I do not get a typeahead suggestion for `--osx`. But when I type
```
tldr --o<tab>
```
it does autocomplete the option to `--osx`. So, it is working partway.

To compare with a different option that does work fully, if I type
```
tldr --v
```
I see a typeahead suggestion that fills in `--version`. And if I hit `<tab>` it does autocomplete `--version`.

The function you've used here is `__tldr_no_os_choice_opt_and_p`. But aren't they making an "os choice" if they have included one of the `--<os>` options? Maybe I don't know exactly what `__tldr_no_os_choice_opt_and_p` does.

Reply:
Ah, yes, I see what you're saying. My mistake!

Reply:
`__tldr_no_os_choice_opt_and_p` is a function that is identical to `-n '__fish_not_contain_opt      linux osx sunos; and __fish_not_contain_opt -s r render; and __tldr_not_contain_standalone_opt; and __fish_not_contain_opt -s p platform'`, where `__tldr_not_contain_standalone_opt` checks to see if any standalone options like `--help` are present. It exists because the check for `--platform` cannot contain a check for itself (otherwise the option completion breaks), while the `--linux`/`--osx`/`--sunos` options should check for the existence of `--platform` (that's what the `_and_p` part means).

It could possibly be given a better name.

Reply:
The part that I find confusing—not necessarily wrong, though—is that the conditions for all the `--<os>` options include a clause for `__fish_not_contain_opt      linux osx sunos`. Isn't that contradictory?

Reply:
That's so they don't get suggested after they have been used. Both `tldr --linux -osx` and `tldr --linux --linux` are invalid options, and without that line both of these outcomes would be completed.

With that line, however, once you type one of the options, like `tldr --sunos`, it will stop suggesting any more options.

This isn't the only option that works like this. Most other options exclude themselves so that they don't get suggested once they've already been typed.

Hopefully that explanation clears it up. Fish completion is very complicated.

Reply:
I will say that it seems plausible, I'm still confused, but I can trust you that this is how it is supposed to work!

Reply:
Fish completion is kind of weird. If you're still confused, I'd suggest that you remove the line and see what changes.